### PR TITLE
Refactor of CPU list handling

### PR DESCRIPTION
**Pull request by mbulatova**
SetAffinity now returns error.
ParseCPU now does only parsing. Add DropInvalidCPUs.
Common test now does not not require many cores.

### CODE REVIEW

File: common/common_test.go [hunk 1]
```diff
@@ -18,41 +18,49 @@ var invalidCpuRangeErr = common.WrapWithNFError(nil, "CPU range is invalid, min
...
+	{"10-14,11-12", []uint{10, 11, 12, 13, 14}, nil},
+	{"1-3,6-", []uint{1, 2, 3}, &strconv.NumError{"Atoi", "", strconv.ErrSyntax}},
+	{"-1", []uint{}, &strconv.NumError{"Atoi", "", strconv.ErrSyntax}},
+	{"10-6", []uint{}, common.GetNFError(invalidCpuRangeErr)},
+	{"1-3,10-6", []uint{1, 2, 3}, common.GetNFError(invalidCpuRangeErr)},
 }
 
-// TestParseCPUs require minumum 22 cores to pass without error.
-// If machine has less cores, test fail with ErrMaxCPUExceed error
 func TestParseCPUs(t *testing.T) {
```
Comment (dkolistratova) [inline common/common_test.go]: I think golint may swear at exported function with no comment.

File: common/common_test.go [hunk 2]
```diff
@@ -18,41 +18,49 @@ var invalidCpuRangeErr = common.WrapWithNFError(nil, "CPU range is invalid, min
...
+var dropInvalidCPUsTests = []struct {
+	cpus     []uint
+	maxcpu   uint
+	expected []uint // expected valid cpu list
+}{
+	{[]uint{}, 20, []uint{}},
+	{[]uint{1, 2, 100, 1, 2, 100}, 20, []uint{1, 2, 1, 2}},
+}
+
+// TestParseCPUs require minumum 22 cores to pass without error.
```
Comment (dkolistratova) [inline common/common_test.go]: TestDropInvalidCPUs

File: flow/flow.go [hunk 3]
```diff
@@ -358,9 +358,10 @@ func SystemInit(args *Config) error {
 	var cpus []uint
 	var err error
 	if args.CPUList != "" {
-		if cpus, err = common.ParseCPUs(args.CPUList, CPUCoresNumber); err != nil {
+		if cpus, err = common.ParseCPUs(args.CPUList); err != nil {
 			return err
 		}
+		cpus = common.DropInvalidCPUs(cpus, CPUCoresNumber)
```
Comment (dkolistratova) [inline flow/flow.go]: Please ask Ilya or Gregory about CPUCoresNumber as max cpu number?

File: scheduler/scheduler.go [hunk 4]
```diff
@@ -156,15 +156,16 @@ func NewScheduler(cpus []uint, schedulerOff bool, schedulerOffRemove bool,
...
 	err = common.WrapWithNFError(nil, "Requested number of cores isn't enough. System needs at least one core per each Set function (except Merger and Stopper) plus one additional core.", common.NotEnoughCores)
 	index := scheduler.getCoreIndex()
 	if index == -1 {
 		return err
 	}
 	core := scheduler.cores[index].id
 	common.LogDebug(common.Initialization, "Start SCHEDULER at", core, "core")
-	low.SetAffinity(uint8(core))
+	if err := low.SetAffinity(uint8(core)); err != nil {
+		return err
```
Comment (dkolistratova) [inline scheduler/scheduler.go]: I think we may wrap it with common.WrapWithNFError

File: scheduler/scheduler.go [hunk 5]
```diff
@@ -176,7 +177,10 @@ func (scheduler *Scheduler) SystemStart() (err error) {
 		common.LogDebug(common.Initialization, "Start STOP at scheduler", core, "core")
 	}
 	go func() {
-		low.SetAffinity(uint8(core))
+		err := low.SetAffinity(uint8(core))
+		if err != nil {
+			common.LogFatal(common.Initialization, "Failed to set affinity to", core, "core: ", err)
```
Comment (dkolistratova) [inline scheduler/scheduler.go]: maybe return common.WrapWithNFError instead of fatal logging?

Comment (dkolistratova) [reply scheduler/scheduler.go]: here and below

Comment (mbulatova) [reply scheduler/scheduler.go]: I place LogFatal only in goroutines, because we do not return values from goroutines

Comment (mbulatova) [reply scheduler/scheduler.go]: Here are several goroutines, each can give this error, maybe we should think of writing its errors to global channel or err variable?

File: common/common_test.go (hunk 1 above)
Comment (ifilippov) [reply common/common_test.go]: No, this is just test function in test file

File: scheduler/scheduler.go (hunk 5 above)
Comment (ifilippov) [reply scheduler/scheduler.go]: LogFatal only in goroutines -> it isn't right logic. Right logic is only via execution. Here is not execution - it is initialization of stop goroutine. So you can return error (one more time our errors are awful) with comment about "for stopping thread"

File: low/low.go [hunk 6]
```diff
@@ -507,14 +507,18 @@ func CreateMempool() *Mempool {
...
 	runtime.LockOSThread()
 
 	var cpuset C.cpu_set_t
 	C.initCPUSet(C.uint8_t(coreID), &cpuset)
-	syscall.RawSyscall(syscall.SYS_SCHED_SETAFFINITY, uintptr(0), unsafe.Sizeof(cpuset), uintptr(unsafe.Pointer(&cpuset)))
+	_, _, errno := syscall.RawSyscall(syscall.SYS_SCHED_SETAFFINITY, uintptr(0), unsafe.Sizeof(cpuset), uintptr(unsafe.Pointer(&cpuset)))
+	if errno != 0 {
+		return common.WrapWithNFError(nil, errno.Error(), common.SetAffinityErr)
+	}
+	return nil
```
Comment (ifilippov) [inline low/low.go]: Was this tested?

File: common/common.go [hunk 7]
```diff
@@ -382,3 +371,16 @@ func removeDuplicates(array []uint) []uint {
...
+func DropInvalidCPUs(nums []uint, maxcpu uint) []uint {
+	i := 0
+	for _, x := range nums {
+		if x < maxcpu {
+			nums[i] = x
+			i++
+		}
+	}
+	return nums[:i]
+}
```
Comment (ifilippov) [inline common/common.go]: Really? Without any notice to user?

**Comment by ifilippov** [comment]
Small refactor? Really? Ideal name

File: low/low.go (hunk 6 above)
Comment (mbulatova) [reply low/low.go]: Yes, I made it to fail and it fails as expected

File: flow/flow.go (hunk 3 above)
Comment (ifilippov) [reply flow/flow.go]: NumCPU - OK, however I suppose that all user input should be handled in one function. Parsing, duplication, trunking, any other points.

File: scheduler/scheduler.go (hunk 5 above)
Comment (mbulatova) [reply scheduler/scheduler.go]: So we need to return value from running goroutine. It seems correct way is to use channel https://www.atatus.com/blog/goroutines-error-handling/

**Comment by ifilippov** [comment]
Approved. Name it "Refactor of CPU list handling"

**Pull request merged**